javalib/src/main/scala/java/util/AbstractList.scala

```diff
@@ -1,3 +1,284 @@
+  }
+
+  protected[this] def checkIndexOnBounds(index: Int): Unit = {
+    if (index < 0 || index > size)
+      throw new IndexOutOfBoundsException(index.toString)
+  }
+
+  private abstract class AbstractListView(fromIndex: Int,
+        private var toIndex: Int) extends AbstractList[E] {
+
+    override def add(e: E): Boolean = {
+      add(size, e)
+      toIndex += 1
+      true
+    }
+
+    override def add(index: Int, e: E): Unit = {
+      checkIndexOnBounds(index)
+      self.add(fromIndex + index, e)
+    }
+
+    override def addAll(index: Int, c: Collection[_ <: E]): Boolean = {
+      checkIndexOnBounds(index)
+      self.addAll(fromIndex + index, c)
+      val elementsAdded = c.size
+      toIndex += elementsAdded
+      elementsAdded != 0
+    }
+
+    override def addAll(c: Collection[_ <: E]): Boolean =
+      addAll(size, c)
+
+    override def clear(): Unit = {
+      self.removeRange(fromIndex, toIndex)
+      toIndex = fromIndex
+    }
+
+    def get(index: Int): E = {
+      checkIndexInBounds(index)
+      self.get(fromIndex + index)
+    }
+
+    override def remove(index: Int): E = {
+      checkIndexInBounds(index)
+      val elem = self.remove(fromIndex + index)
+      toIndex -= 1
+      elem
+    }
+
+    override def set(index: Int, e: E): E = {
+      checkIndexInBounds(index)
+      self.set(fromIndex + index, e)
+    }
+
+    def size(): Int =
+      toIndex - fromIndex
+
+  }
+
+  /*
+   * RandomAccessListIterator implementation assumes that the has an efficient
+   * .get(index) implementation.
+   */
+  private class RandomAccessListIterator(private var i: Int, start: Int,
+        private var end: Int) extends ListIterator[E] {
+
+    private var last = -1
+
+    def hasNext(): Boolean =
+      i < end
+
+    def next(): E = {
+      last = i
+      i += 1
+      get(last)
+    }
+
+    def hasPrevious(): Boolean =
+      start < i
+
+    def previous(): E = {
+      last = i
+      i -= 1
+      get(i)
+    }
+
+    def nextIndex(): Int = i
+
+    def previousIndex(): Int = i - 1
+
+    def remove(): Unit = {
+      if (i == -1 || i == last)
+        throw new IllegalStateException
+
+      self.remove(last)
+      if (last < i)
+        i -= 1
+      last = i
+      end -= 1
+    }
+
+    def set(e: E): Unit = {
+      if (last == -1)
+        throw new IllegalStateException
+
+      self.set(last, e)
+    }
+
+    def add(e: E): Unit = {
+      self.add(i, e)
+      end += 1
+      last = i
+      i += 1
+    }
+  }
+
+  /*
+   * BackedUpListIterator implementation assumes that the underling list is not
+   * necessarily on a RandomAccess list. Hence it wraps the underling list
+   * iterator and assumes that this one is more efficient than accessing
+   * elements by index.
+   */
+  private class BackedUpListIterator(private var index: Int, fromIndex: Int) extends ListIterator[E] {
+
+    private val innerIterator = self.listIterator(fromIndex + index)
+
+    private var last = -1
```

Comment:
If we don't need `index`, we don't need `last` either, because `last` is used only to update `index`.